walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,102 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+    // first iterate through and combine events that overlap
+    // subtract combined events calendar from whole day
+    // remove time ranges that are too short 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+        List<TimeRange> avaliableRanges = new ArrayList<TimeRange>();
+        avaliableRanges.add(TimeRange.WHOLE_DAY);
+        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+
+        for (TimeRange booked : bookedRanges) {
+            TimeRange lastInAvaliable = avaliableRanges.get(avaliableRanges.size()-1);
+            if (lastInAvaliable.overlaps(booked)){
+                List<TimeRange> newLastAvaliable = removeOverlap(lastInAvaliable, booked);
+                avaliableRanges.remove(lastInAvaliable);
+                avaliableRanges.addAll(newLastAvaliable);
+                Collections.sort(avaliableRanges, TimeRange.ORDER_BY_START);
+            }
+        }
+
+        // remove TimeRanges too small
+        for (TimeRange avaliable : avaliableRanges) {
+            if (avaliable.duration() < request.getDuration()) {
+                avaliableRanges.remove(avaliable);
+            }
+        }
+
+        return avaliableRanges;
+  }
+
+// returns set of TimeRanges formed by all events that have mandatory attendees from the request with no overlap
+  public Collection<TimeRange> combinedMandatoryRanges(Collection<Event> events, MeetingRequest request) {
+        List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+        for (Event event : events) {
+            if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) ) { //events with mandatory attendees only
+                busyTimes.add(event.getWhen());
+                System.out.println(event.getWhen());
+            }
+        }
+        Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+
+        List<TimeRange> combinedBusyTimes = new ArrayList<TimeRange>();
+
+        int i = 0;
+        int start = -1;
+        int end;
+        while (i < busyTimes.size()) {
```

Comment:
Since you're always incrementing i, it's clearer to write this as a for loop (then your iteration variable i won't be in scope outside of the for loop)

Reply:
noted and implemented